src/kerberos/client/generators.rs

```diff
@@ -345,7 +346,100 @@ pub fn generate_tgs_req(options: GenerateTgsReqOptions) -> Result<TgsReq> {
 #[derive(Debug)]
 pub struct ChecksumOptions {
     pub checksum_type: Vec<u8>,
-    pub checksum_value: Vec<u8>,
+    pub checksum_value: ChecksumValues,
+}
+
+#[derive(Debug)]
+pub struct ChecksumValues {
+    inner: Vec<u8>, // use named fields for future extensibility, this is a temporary solution
```

Comment:
suggestion: Use `[u8; 24]` instead of `Vec<u8>`.

```diff
diff --git a/src/kerberos/client/generators.rs b/src/kerberos/client/generators.rs
index d1b31df..cfcb1b7 100644
--- a/src/kerberos/client/generators.rs
+++ b/src/kerberos/client/generators.rs
@@ -351,18 +351,18 @@ pub struct ChecksumOptions {
 
 #[derive(Debug)]
 pub struct ChecksumValues {
-    inner: Vec<u8>, // use named fields for future extensibility, this is a temporary solution
+    inner: [u8; 24], // use named fields for future extensibility, this is a temporary solution
 }
 
 impl Default for ChecksumValues {
     fn default() -> Self {
         Self {
-            inner: AUTHENTICATOR_DEFAULT_CHECKSUM.to_vec(),
+            inner: AUTHENTICATOR_DEFAULT_CHECKSUM,
         }
     }
 }
 
-impl From<ChecksumValues> for Vec<u8> {
+impl From<ChecksumValues> for [u8; 24] {
     fn from(val: ChecksumValues) -> Self {
         val.inner
     }
@@ -370,9 +370,7 @@ impl From<ChecksumValues> for Vec<u8> {
 
 impl From<[u8; 24]> for ChecksumValues {
     fn from(bytes: [u8; 24]) -> Self {
-        ChecksumValues {
-            inner: Vec::from(bytes),
-        }
+        ChecksumValues { inner: bytes }
     }
 }
 
@@ -383,7 +381,7 @@ impl ChecksumValues {
         self.inner[20..24].copy_from_slice(&flag_bytes);
     }
 
-    pub(crate) fn get_inner(self) -> Vec<u8> {
+    pub(crate) fn get_inner(self) -> [u8; 24] {
         self.inner
     }
 }
@@ -511,7 +509,7 @@ pub fn generate_authenticator(options: GenerateAuthenticatorOptions) -> Result<A
         }
         Optional::from(Some(ExplicitContextTag3::from(Checksum {
             cksumtype: ExplicitContextTag0::from(IntegerAsn1::from(checksum_type)),
-            checksum: ExplicitContextTag1::from(OctetStringAsn1::from(checksum_value)),
+            checksum: ExplicitContextTag1::from(OctetStringAsn1::from(checksum_value.to_vec())),
         })))
     } else {
         Optional::from(None)
diff --git a/src/pku2u/generators.rs b/src/pku2u/generators.rs
index a0113b0..ea9864b 100644
--- a/src/pku2u/generators.rs
+++ b/src/pku2u/generators.rs
@@ -254,7 +254,7 @@ pub fn generate_authenticator(options: GenerateAuthenticatorOptions) -> Result<A
         checksum_value,
     }) = checksum
     {
-        let mut checksum_value = checksum_value.get_inner();
+        let mut checksum_value = checksum_value.get_inner().to_vec();
         if checksum_type == AUTHENTICATOR_CHECKSUM_TYPE && channel_bindings.is_some() {
             if checksum_value.len() < 20 {
                 return Err(Error::new(
```

rationale: 1) Here, a checksum is always 24 bytes. Therefore, `[u8; 24]` is just the right type. 2) `Vec<u8>` is itself taking 24 bytes, excluding the 24 bytes of actual payload on the heap. `[u8; 24]` is 24 bytes by itself, payload only, no allocation. ⇒ There is no benefit in using `Vec<u8>`

Reply:
The reason why I use Vec<u8> here is because it is actually not fixed to 24 bytes. According to the RFC 4121, 4.1.1, there are optional values that could be added in the future.